webhook/aws/sns_handler.go

```diff
@@ -0,0 +1,243 @@
+		Token:    aws.String(msg.Token),    // Required
+		TopicArn: aws.String(msg.TopicArn), // Required
+	}
+	resp, err := ss.SVC.ConfirmSubscription(params)
+	if err != nil {
+		log.Error("SNS confirm error %v", err)
+		// TODO return error response
+		return
+	}
+
+	log.Debug("SNS confirm response: %v", resp)
+	
+	// Thread synchronization acquire lock and update
+	ss.UpdateSubscriptionArn(resp.SubscriptionArn)
+
+	ResponseJson(`{"message":"ok"}`, rw)
+
+}
+
+// Decodes SNS Notification message and returns 
+// the actual message which is json webhook content
+func (ss *SNSServer) NotificationHandle(rw http.ResponseWriter, req *http.Request) []byte {
+	log := ss.logger()
+	defer func() {
+		if r := recover(); r != nil {
+			log.Error( "SNS NotificationHandle recover error %v", r )
+		}
+	}()
+	subArn := req.Header.Get("X-Amz-Sns-Subscription-Arn")
+	if !ss.ValidateSubscriptionArn(subArn) {
+		ResponseJsonErr(rw, "SubscriptionARN not match", http.StatusBadRequest)
+		return nil
+	}
+	msg := new(SNSMessage)
+
+	raw, err := DecodeJsonPayload(req, msg)
+	if err != nil {
+		log.Error("SNS read req body error %v", err)
+		ResponseJsonErr(rw, "request body error", http.StatusBadRequest)
+		return nil
+	}
+	//health.SendEvent(HTH.Set("TotalDataPayloadReceived", int(len(raw)) ))
+	
+	log.Debug("SNS notification payload raw [%v]", string(raw))
+	
+	// TODO: Verify SNS Message authenticity by verifying signature
+
+	EnvAttr := msg.MessageAttributes[MSG_ATTR]
+	log.Trace("SQS notification EnvAttr %v", EnvAttr)
+
+	msgEnv := EnvAttr.Value
+	log.Trace("SNS notification msgEnv %v",  msgEnv)
+	if msgEnv != ss.Config.Env {
+		log.Warn("SNS msg env %v does not match config env %v", msgEnv, ss.Config.Env)
+		ResponseJsonErr(rw, "request body error", http.StatusNotAcceptable)
+		return nil
+	}
+	
+	return []byte(msg.Message)
+}
+
+// Publish Notification message to AWS SNS topic 
+// @return status 
+// 1. If SNS Server is not yet ready then return status as false 
+// to indicate that Publish was not successful
+// 2. If Publish to AWS SNS fails then return false
+// Retry after some time when the status returned from this function is false
+func (ss *SNSServer) PublishMessage(message string) bool {
+	log := ss.logger()
+	log.Debug("SNS PublishMessage called %v ", message)
+	defer func() {
+		if r := recover(); r != nil {
+			log.Error( "SNS PublishMessage recover error %v", r )
+		}
+	}()
+
+	// check if ready to accept messages before publishing message.
+	if !ss.IsReady() {
+		return false
+	}
+
+	params := &sns.PublishInput{
+		Message: aws.String(message), // Required
+		MessageAttributes: map[string]*sns.MessageAttributeValue{
+			MSG_ATTR: { // Required
+				DataType:    aws.String("String"), // Required
+				StringValue: aws.String(ss.Config.Env),
+			},
+		},
+		MessageStructure: aws.String("messageStructure"),
+		Subject:          aws.String("subject new"),
+		TopicArn:         aws.String(ss.Config.Sns.TopicArn),
+	}
+	resp, err := ss.SVC.Publish(params)
```

Comment:
.... continuing from above:  here you would do `somechannel <- params` and then you're done.  You would then need a goroutine running that looked roughly like:

```
func publish(svc *sns.SNS, input <-chan *sns.PublishInput, output chan<- *sns.PublishOutput) {
    // if you want this for loop to exit, just close(input) from another goroutine
    for publishInput := range input {
        response, err := svc.Publish(publishInput)
        // logging as desired

        if err == nil {
            // this would, of course, require another goroutine to consume the output
            // I don't know if you need this, this is just an example
            // you could just log the response and drop it
            output <- response
        }
    }
}
```